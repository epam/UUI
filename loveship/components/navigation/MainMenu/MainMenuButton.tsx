import css from './MainMenuButton.scss';
import { Button, ButtonProps } from '@epam/uui-components';
import { IAdaptiveItem, withMods } from '@epam/uui-core';
import { ReactComponent as TriangleIcon } from '../../icons/chevron-down-24.svg';

export interface MainMenuButtonMods {
    type?: 'primary' | 'secondary';
}

export const MainMenuButton = withMods<ButtonProps, MainMenuButtonMods & IAdaptiveItem>(
    Button,
    mods => [css.root, css['type-' + (mods.type || 'primary')], css.fontSansSemibold],
    () => ({ dropdownIcon: TriangleIcon })
);
