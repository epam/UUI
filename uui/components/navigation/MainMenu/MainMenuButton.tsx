import * as React from 'react';
import {
    IAdaptiveItem,
    ICanBeActive,
    Icon,
    IDropdownToggler,
    IHasCaption,
    IHasIcon,
    useIsActive,
    uuiElement, uuiMod,
} from '@epam/uui-core';
import { Clickable, ClickableComponentProps, IconContainer } from '@epam/uui-components';
import { CountIndicator } from '../../widgets/CountIndicator';
import { ReactComponent as SvgTriangle } from '@epam/assets/icons/navigation-chevron_down-outline.svg';
import css from './MainMenuButton.module.scss';

interface MainMenuButtonMods {
    /**
    * Defines component type. The primary button leads to the main pages of the site, and the secondary to the others.
    */
    type?: 'primary' | 'secondary';
}

export type MainMenuButtonProps = MainMenuButtonMods & IAdaptiveItem & IDropdownToggler & Omit<ClickableComponentProps, 'isDisabled'>
& IHasIcon & IHasCaption & ICanBeActive & {
    /** Icon for drop-down toggler */
    dropdownIcon?: Icon;
    /** Count value to be placed in component */
    count?: React.ReactNode;
};

export const MainMenuButton = React.forwardRef<HTMLButtonElement | HTMLAnchorElement | HTMLSpanElement, MainMenuButtonProps>((props, ref) => {
    const { type, ...clickableProps } = props;

    const { isActive } = useIsActive({
        isLinkActive: props.isLinkActive,
        link: props.link,
        isActive: props.isActive,
    });

    return (
        <Clickable
            { ...clickableProps }
            rawProps={ {
                'aria-expanded': props.isOpen,
                'aria-current': isActive
                    ? 'page'
                    : undefined,
                ...props.rawProps,
            } }
            cx={ [
                css.root,
                css['type-' + (type || 'primary')],
                isActive && uuiMod.active,
                props.cx,
            ] }
            ref={ ref }
        >
            { props.icon && props.iconPosition !== 'right' && (
                <IconContainer
                    icon={ props.icon }
                    onClick={ props.onIconClick }
                />
            ) }
            { props.caption && (
                <div className={ uuiElement.caption }>
                    { props.caption }
                </div>
            ) }
            { props.count !== undefined && props.count !== null && (
                <CountIndicator
                    caption={ props.count }
                    color="neutral"
                    size="18"
                />
            ) }
            { props.icon && props.iconPosition === 'right' && (
                <IconContainer icon={ props.icon } onClick={ props.onIconClick } />
            ) }
            { props.isDropdown && (
                <IconContainer icon={ SvgTriangle } flipY={ props.isOpen } />
            )}
        </Clickable>
    );
});
