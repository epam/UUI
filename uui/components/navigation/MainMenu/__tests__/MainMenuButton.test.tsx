import React from 'react';
import { renderSnapshotWithContextAsync } from '@epam/uui-test-utils';
import { MainMenuButton } from '../MainMenuButton';

describe('MainMenuButton', () => {
    it('should be rendered correctly with minimum props', async () => {
        const tree = await renderSnapshotWithContextAsync(<MainMenuButton />);
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <MainMenuButton
                caption="Test button"
                type="primary"
                target="_blank"
                link={ { pathname: '/' } }
                isActive
                isDropdown
                isOpen={ false }
                collapseToMore
                estimatedWidth={ 120 }
                showInBurgerMenu
                priority={ 6 }
            />,
        );
        expect(tree).toMatchSnapshot();
    });
});
