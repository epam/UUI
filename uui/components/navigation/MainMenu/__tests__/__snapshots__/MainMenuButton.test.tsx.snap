// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainMenuButton should be rendered correctly 1`] = `
<DocumentFragment>
  <a
    aria-current="page"
    aria-expanded="false"
    class="uui-button-box uui-enabled -clickable uui-anchor root type-primary uui-active"
    href=""
    rel="noopener noreferrer"
    tabindex="0"
    target="_blank"
  >
    <div
      class="uui-caption"
    >
      Test button
    </div>
    <div
      class="container uui-icon uui-enabled"
    >
      <svg
        class=""
      />
    </div>
  </a>
</DocumentFragment>
`;

exports[`MainMenuButton should be rendered correctly with minimum props 1`] = `
<DocumentFragment>
  <span
    class="uui-button-box uui-enabled root type-primary"
    tabindex="-1"
  />
</DocumentFragment>
`;
